src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java

```diff
@@ -71,13 +79,88 @@ public Set<String> supportedFileAttributeViews() {
         return SupportedFileFileAttributeViewsHolder.supportedFileAttributeViews;
     }
 
+    /**
+     * Clones the file whose path name is {@code src} to that whose path
+     * name is {@code dst} using the {@code clonefile} system call.
+     *
+     * @param src the path of the source file
+     * @param dst the path of the destination file (clone)
+     * @param followLinks whether to follow links
+     *
+     * @return 0 on success, IOStatus.UNSUPPORTED_CASE if the call does not work
+     *         with the given parameters, or IOStatus.UNSUPPORTED if cloning is
+     *         not supported on this platform
+     */
+    private int clone(UnixPath src, UnixPath dst, boolean followLinks)
+        throws IOException
+    {
+        // Do not attempt cloning if the source volume does not support it
+        long options = followLinks ? 0 : FSOPT_NOFOLLOW;
+        if ((BsdNativeDispatcher.getattrlist(src, options) &
+            VOL_CAP_INT_CLONE) == 0)
+            return IOStatus.UNSUPPORTED_CASE;
+
+        int flags = followLinks ? 0 : CLONE_NOFOLLOW;
+        try {
+            BsdNativeDispatcher.clonefile(src, dst, flags);
+        } catch (UnixException x) {
+            switch (x.errno()) {
+                case ENOTSUP: // cloning not supported by filesystem
+                    return IOStatus.UNSUPPORTED;
+                case EXDEV:   // src and dst on different filesystems
+                case ENOTDIR: // problematic path parameter(s)
+                    return IOStatus.UNSUPPORTED_CASE;
+                default:
+                    x.rethrowAsIOException(src, dst);
+                    return IOStatus.THROWN;
+            }
+        }
+
+        return 0;
+    }
+
     @Override
     protected int directCopy(int dst, int src, long addressToPollForCancel)
         throws UnixException
     {
         return directCopy0(dst, src, addressToPollForCancel);
     }
 
+    @Override
+    protected void copyFile(UnixPath source,
+                            UnixFileAttributes attrs,
+                            UnixPath target,
+                            Flags flags,
+                            long addressToPollForCancel)
+        throws IOException
+    {
+        // Attempt to clone the source unless cloning is not supported,
+        // cancellation is not possible, or attributes are not to be copied
+        if (!cloneFileNotSupported && addressToPollForCancel == 0 &&
+            flags.copyPosixAttributes) {
+            int res = clone(source, target, flags.followLinks);
+
+            if (res == 0) {
+                // copy owner (not done by clonefile)
+                try {
+                    chown(target, attrs.uid(), attrs.gid());
+                } catch (UnixException x) {
+                    if (flags.failIfUnableToCopyPosix)
+                        x.rethrowAsIOException(target);
```

Comment:
Can you compare this to the Unix implementation as I thought there was a best attempt to unlink when throwing an exception?

Reply:
You are correct: thanks for catching my gaffe. Fixed in f51cd186fe624d03ce7c6a4f49140998e2d48322.